src/f32-gemm/MrxLmulv-rvv.c.in

```diff
@@ -0,0 +1,170 @@
+// Copyright 2023 SiFive, Inc.
+//
+// This source code is licensed under the BSD-style license found in the
+// LICENSE file in the root directory of this source tree.
+
+$assert DATATYPE in ["F32"]
+$assert ACTIVATION in ["LINEAR", "RELU", "MINMAX"]
+$assert MR >= 1
+$assert LMUL in [1, 2, 4]
+#include <assert.h>
+
+#include <riscv_vector.h>
+
+#include <xnnpack/gemm.h>
+
+
+$DATATYPE_SPEC = {"F32": "f32"}[DATATYPE]
+$SUFFIX = {"LINEAR": "", "RELU": "_relu", "MINMAX": "_minmax"}[ACTIVATION]
+$PARAMS = {"LINEAR": "xnn_f32_default_params", "RELU": "xnn_f32_relu_params", "MINMAX": "xnn_f32_minmax_params"}[ACTIVATION]
+void xnn_${DATATYPE_SPEC}_gemm${SUFFIX}_ukernel_${MR}x${LMUL}v__rvv(
+    size_t mr,
+    size_t nc,
+    size_t kc,
+    const float* restrict a,
+    size_t a_stride,
+    const float* restrict w,
+    float* restrict c,
+    size_t cm_stride,
+    size_t cn_stride,
+    const union ${PARAMS} params[restrict XNN_MIN_ELEMENTS(1)])
+{
+  assert(mr != 0);
+  assert(mr <= ${MR});
+  assert(nc != 0);
+  assert(kc != 0);
+  assert(kc % sizeof(float) == 0);
+  assert(a != NULL);
+  assert(w != NULL);
+  assert(c != NULL);
+
+  $if ACTIVATION == "MINMAX":
+    const float vmin = params->scalar.min;
+    const float vmax = params->scalar.max;
+  const float* a0 = a;
+  float* c0 = c;
+  $if MR > 1:
+    const float* a1 = (const float*) ((uintptr_t) a + a_stride);
+    float* c1 = (float*) ((uintptr_t) c + cm_stride);
+    $for M in range(2, MR):
+      const float* a${M} = (const float*) ((uintptr_t) a${M-1} + a_stride);
+      float* c${M} = (float*) ((uintptr_t) c${M-1} + cm_stride);
+
+  const size_t nr = __riscv_vsetvlmax_e32m${LMUL}();
+  size_t vl = nr;
+  if XNN_LIKELY(mr == ${MR}) {
```

Comment:
remove full MR check and clamp pointers like arm/intel kernels.  There will be redundent loads/math but we have a specialized MR=1 kernel and can do multiple kernels if small MR comes up much.


const float* a0 = a;
float* c0 = c;
$for M in range(1, MR):
  const float* a${M} = (const float*) ((uintptr_t) a${M-1} + a_stride);
  float* c${M} = (float*) ((uintptr_t) c${M-1} + cm_stride);
  $if M % 2 == 0:
    if XNN_UNPREDICTABLE(mr <= ${M}) {
      a${M} = a${M-1};
      c${M} = c${M-1};
    }
  $elif M + 1 == MR:
    if XNN_UNPREDICTABLE(mr != ${M+1}) {
      a${M} = a${M-1};
      c${M} = c${M-1};
    }
  $else:
    if XNN_UNPREDICTABLE(mr < ${M+1}) {
      a${M} = a${M-1};
      c${M} = c${M-1};
    }

Reply:
OK, I'll follow ARM & Intel implementation.